From 17643fd28f499691ae97503eb58755b4b9fabb53 Mon Sep 17 00:00:00 2001
From: Nick Schermer <nick@xfce.org>
Date: Sun, 22 Jul 2012 09:43:51 +0000
Subject: Actions: Fix logic of session saving (bug #8857).

---
diff --git a/plugins/actions/actions.c b/plugins/actions/actions.c
index ebc7b5b..53ff8d0 100644
--- a/plugins/actions/actions.c
+++ b/plugins/actions/actions.c
@@ -705,7 +705,7 @@ actions_plugin_action_confirmation_time (gpointer data)
 
   if (timeout->time_left == 0)
     {
-      /* unattended shutdown, so don't allow apps to cancel shutdown */
+      /* unattended shutdown, don't save the session to avoid blocking the logout */
       timeout->unattended = TRUE;
 
       gtk_dialog_response (GTK_DIALOG (timeout->dialog),
@@ -928,22 +928,22 @@ actions_plugin_action_activate (GtkWidget      *widget,
     {
     case ACTION_TYPE_LOGOUT:
       succeed = actions_plugin_action_dbus_xfsm ("Logout", FALSE,
-                                                 unattended, &error);
+                                                 !unattended, &error);
       break;
 
     case ACTION_TYPE_LOGOUT_DIALOG:
       succeed = actions_plugin_action_dbus_xfsm ("Logout", TRUE,
-                                                 unattended, &error);
+                                                 !unattended, &error);
       break;
 
     case ACTION_TYPE_RESTART:
       succeed = actions_plugin_action_dbus_xfsm ("Restart", FALSE,
-                                                 unattended, &error);
+                                                 !unattended, &error);
       break;
 
     case ACTION_TYPE_SHUTDOWN:
       succeed = actions_plugin_action_dbus_xfsm ("Shutdown", FALSE,
-                                                 unattended, &error);
+                                                 !unattended, &error);
       break;
 
     case ACTION_TYPE_HIBERNATE:
--
cgit v0.9.0.3
